### DISCUSSION TITLE
[Bug]: Nginx dramatic port proxying ：302 not found

### DISCUSSION THREAD

**Discussion by damonwan1** [coder/code-server#5120] [Q&A]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### OS/Web Information

- Web Browser:   Chrome
- Local OS:   win10
- Remote OS: Centos 7.6
- Remote Architecture: ARM
- `code-server --version`: 4.3.0

### Steps to Reproduce

My requirement is to achieve dramatic port proxy with nginx and code-server. 

For Example，there is a situation.
1. one user access our website and click "create my vscode"button, it will run code-server by "code-server --auth none --host 0.0.0.0 --port 20000", so remote server A has started a code-server service on port 20000. (**Different user has different port number20001,20002,etc.**)
2. User access the code-server website by url "http://mywebsite.com/vscode/20000/" 
3. The remote  application server B has deployed nginx, the user request will firstly target serverB and the nginx on serverB proxy the request to server A(code-server). sry, it is a simply process.
4. The nginx conf is below.  I have tried all the method of conf, but it didn't work.
```
                location ~* /vscode/([0-9]+)/
                {
                         proxy_pass  http://137.2.2.1:$1/;
                         proxy_set_header X-Real-IP $remote_addr;
                         proxy_set_header Host $host;
                         proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
                         proxy_http_version 1.1;
                         proxy_set_header Upgrade $http_upgrade;
                         proxy_set_header Connection "upgrade";
                         proxy_redirect off;
                }
```
It didn't work ,too.

```
              location ^~/vscode/20000/
                {
                         set $ter 20000
                         proxy_pass  http://137.2.2.1:$ter/;
                         proxy_http_version 1.1;
                         proxy_set_header Upgrade $http_upgrade;
                         proxy_set_header Connection "Upgrade";
                }
```

However, the non-dramatic params work well.

```
           location ^~/vscode/20000/
                {
                         proxy_pass  http://122.10.10.112:20000/;
                         proxy_http_version 1.1;
                         proxy_set_header Upgrade $http_upgrade;
                         proxy_set_header Connection "Upgrade";
                }
```

### Expected

Actually, i really need a dramatic port proxy, by http://mywebsite.com/vscode/20001/, http://mywebsite.com/vscode/20002/,http://mywebsite.com/vscode/20003/  automaticly.### 

### Actual

The wrong proxying process throws 302 not found.

I checked it by chrome F12, it could find a lot of 302 redirect request. by the way, i have flush my coockies.

### Screenshot/Video

![image](https://user-images.githubusercontent.com/33806362/164444306-d03d277b-7008-49b8-8fa5-92d2b4437649.png)

### Does this issue happen in VS Code?

- [X] I cannot reproduce this in VS Code.

### Are you accessing code-server over HTTPS?

- [X] I am using HTTPS.

### Notes

Thanks a lot for solving this problem. we have tried so many ways and viewed a lot of issues, googles...

**Comment by jsjoeio** [comment]
Thanks for details! Could you provide steps to create a minimal reproduction? I haven't used nginx but happy to try to reproduce this locally. I'm going to transfer this to a Discussion instead.

**Comment by damonwan1** [comment]
1. Install nginx on Server A
2. The nginx.conf file on Server A is as below, start the nginx with the conf file.
```
user  root;
worker_processes  2;

error_log  logs/error.log;
error_log  logs/error.log  notice;
error_log  logs/error.log  info;

events {
    worker_connections  1024;
}


http {
    include       mime.types;
    default_type  application/octet-stream;
    keepalive_timeout  6000s;
    client_max_body_size  20000m;
    proxy_max_temp_file_size 0;
    proxy_buffering off;
    send_timeout 1200s;
    proxy_read_timeout 120s;

    server {
		listen       80 default_server;
		server_name  localhost;
		index index.html index.htm index.php;
		 
        location ^~ /vscode/20000/
        {
            proxy_pass  http://127.0.0.1:20000/;
            proxy_http_version 1.1;
            proxy_set_header Upgrade $http_upgrade;
            proxy_set_header Connection "Upgrade";
        } 
		
      error_page   500 502 503 504  /50x.html;
      location = /50x.html {
          root   html;
      }
    }
}
```
3. ssh port forwarding:  `$ ssh -L <local-port-to-listen>:<remote-host>:<remote-port> <sshserver>
`
Specifically， `ssh  -o StrictHostKeyChecking=no  -L 20000:137.2.2.2:20000 user@sshserver `

4. On Server B（137.2.2.2），start code-server service.
`  code-server --auth none --host 0.0.0.0 --port 20000`

5. Now you can access the port 20000 of server A to use code server.

**Reply by damonwan1** [reply]
1. If i change the nginx.conf to nginx2.conf, it doesn't work.
```
user  root;
worker_processes  2;

error_log  logs/error.log;
error_log  logs/error.log  notice;
error_log  logs/error.log  info;

events {
    worker_connections  1024;
}


http {
    include       mime.types;
    default_type  application/octet-stream;
    keepalive_timeout  6000s;
    client_max_body_size  20000m;
    proxy_max_temp_file_size 0;
    proxy_buffering off;
    send_timeout 1200s;
    proxy_read_timeout 120s;

    server {
		listen       80 default_server;
		server_name  localhost;
		index index.html index.htm index.php;
		 
        location ^~ /vscode/20000/
        {
            set $sigtest 20000;
            proxy_pass  http://127.0.0.1:$sigtest/;
            proxy_http_version 1.1;
            proxy_set_header Upgrade $http_upgrade;
            proxy_set_header Connection "Upgrade";
        } 
		
      error_page   500 502 503 504  /50x.html;
      location = /50x.html {
          root   html;
      }
    }
}
```
2. I tried the conf with Jupyter Lab, it is okay. But not code-server.
3. It seems like the conf1 and conf2 is the same results, but it doesn't.

**Comment by damonwan1** [comment]
1. Actually, what i want is dramatically port forwarding by nginx as nginx3.conf
```
user  root;
worker_processes  2;

error_log  logs/error.log;
error_log  logs/error.log  notice;
error_log  logs/error.log  info;

events {
    worker_connections  1024;
}


http {
    include       mime.types;
    default_type  application/octet-stream;
    keepalive_timeout  6000s;
    client_max_body_size  20000m;
    proxy_max_temp_file_size 0;
    proxy_buffering off;
    send_timeout 1200s;
    proxy_read_timeout 120s;

    server {
		listen       80 default_server;
		server_name  localhost;
		index index.html index.htm index.php;
		 
        location ~^/vscode/([0-9]+){
            	 	proxy_pass  http://127.0.0.1:$1;
                        proxy_set_header X-Real-IP $remote_addr;
                        proxy_set_header Host $host;
                        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
                        proxy_http_version 1.1;
                        proxy_set_header Upgrade $http_upgrade;
                        proxy_set_header Connection "upgrade";
                        proxy_redirect off;
		} 
		
      error_page   500 502 503 504  /50x.html;
      location = /50x.html {
          root   html;
      }
    }
}
```

2. And the code-server on Server B(137.2.2.2) start with changeable params --port.
3. I don't know it is nginx conf problem or code-server problem.  Who can help me , Thanks very much!
4. The nginx3.conf works well on Jupyter Lab.

**Reply by jsjoeio** [reply]
If you remove nginx from the equation, does code-server work?
